core/queryalgebra/evaluation/src/test/java/org/eclipse/rdf4j/query/algebra/evaluation/iterator/GroupIteratorTest.java

```diff
@@ -260,6 +251,46 @@ public void testCustomAggregateFunction_WrongIri() throws QueryEvaluationExcepti
 		}
 	}
 
+	@Test
+	public void testGroupIteratorClose() throws QueryEvaluationException, InterruptedException {
+		// Lock which is already locked to block the thread driving the iteration
+		Lock lock = new ReentrantLock();
+		lock.lock();
+		// Latch to rendezvous on with the iterating thread
+		CountDownLatch iterating = new CountDownLatch(1);
+		// Latch to record whether the iteration under GroupIterator was closed
+		CountDownLatch closed = new CountDownLatch(1);
+
+		EvaluationStrategy evaluator = new StrictEvaluationStrategy(null, null) {
+			@Override
+			protected QueryEvaluationStep prepare(EmptySet emptySet, QueryEvaluationContext context)
+					throws QueryEvaluationException {
+				return bindings -> new LookAheadIteration<>() {
+					@Override
+					protected BindingSet getNextElement() {
+						iterating.countDown(); // signal to test thread iteration started
+						lock.lock(); // block iterating thread
+						return null;
+					}
+
+					@Override
+					protected void handleClose() {
+						closed.countDown();
+					}
+				};
+			}
+		};
+
+		Group group = new Group(new EmptySet());
+		GroupIterator groupIterator = new GroupIterator(evaluator, group, EmptyBindingSet.getInstance(), context);
+
+		Thread iteratorThread = new Thread(groupIterator::next);
+		iteratorThread.start();
+		assertThat(iterating.await(5, TimeUnit.SECONDS)).isTrue();
+		groupIterator.close();
+		assertThat(closed.await(5, TimeUnit.SECONDS)).isTrue();
```

Comment:
Can we also assert that the iteratorThread is no longer alive? So we are sure that we're not leaking a thread.

Reply:
The test _was_ in fact leaking a Thread. Just fixed it.